### DISCUSSION TITLE
每过一段时间，代理会自动失效。及我在使用过程中遇到的几个小问题

### DISCUSSION THREAD

**Discussion by wizard-leoliao** [Openwrt-Passwall/openwrt-passwall2#347] [Q&A]
我的硬件情况：n6000+i226
固件版本：OpenWrt R23.7.12 (2023-07-19) / LuCI Master (git-23.192.03985-da0a596)
内核版本：6.1.38
由passwall2分流，远端dns useipv4+cloudflare doh；本地dns使用adguardhome，并将5553重定向设置为“作为dnsmasq的上行服务器”

症状如题，每隔大概5-6个小时，所有代理都会失效。
具体表现为运行状态栏位百度和github连接正常，谷歌连接失败。所有被代理的ip均无法正常访问谷歌及其他网站。
目前摸索出来的解决办法：清空ipset似乎会临时解决问题，但无法彻底解决，一段时间后会复发。
openwrt的系统日志、passwall2的节点日志均未找到有意义的报错项。
求解原理及解决办法。

顺带一提，我在使用过程中发现有几个不太方便的地方。
1.似乎passwall2保存并应用似乎没办法直接改变passwall2的运行状态。包括但不限于关闭主开关、修改代理规则、修改节点必须在清空一次ipset后才能生效。
2.使用tproxy似乎无法正常访问telegram
3.在规则管理页面内任何一点点的错误(比如geosite在dat文件中不存在、把geosite填到geoip内、或者格式错误)都将导致passwall2工作异常，并使得整个openwrt无法正常工作，解决这个问题我摸索了挺长时间的🤣。
如果能在某个地方提示下，填错规则可能导致全局断网，或者提供一个查询geosite、geoip的栏位，方便自定义规则就好了。

下方为日志
```
2023-09-04 01:24:27: 删除相关防火墙规则完成。
2023-09-04 01:24:30: 清空并关闭相关程序和缓存完成。
2023-09-04 01:24:31: 127.0.0.1#15353 （ 直连DNS：127.0.0.1#5553 远程DNS：https://1.1.1.2/dns-query）
2023-09-04 01:24:31:   - [0]节点列表中的域名(vpslist)：127.0.0.1#5553
2023-09-04 01:24:31:   - [0]默认：127.0.0.1#15353
2023-09-04 01:24:32: 开始加载防火墙规则...
2023-09-04 01:24:32: 加入负载均衡的节点到ipset[passwall2_vpslist]直连完成
2023-09-04 01:24:32: 加入所有节点到ipset[passwall2_vpslist]直连完成
2023-09-04 01:24:32: 加载路由器自身 TCP 代理...
2023-09-04 01:24:32:   - [0]不代理TCP 端口：51413,18443,55555,6580,58769
2023-09-04 01:24:32: 加载路由器自身 UDP 代理...
2023-09-04 01:24:32:   - [0]不代理 UDP 端口：1:65535
2023-09-04 01:24:32: 访问控制：
2023-09-04 01:24:32: 访问控制：
2023-09-04 01:24:32:   - 备注【server】，IP【192.168.11.127】，使用TCP节点[分流总节点] [全局代理](REDIRECT:1041)代理[2]除51413,18443,55555,6580,58769外的所有端口
2023-09-04 01:24:32:   - 备注【server】，IP【192.168.11.127】，使用UDP节点[分流总节点] [全局代理](TPROXY:1041)代理[2]除1:65535外的所有端口
2023-09-04 01:24:32:   - 备注【server】，IP【192.168.11.115】，使用TCP节点[分流总节点] [全局代理](REDIRECT:1041)代理[2]除51413,18443,55555,6580,58769外的所有端口
2023-09-04 01:24:32:   - 备注【server】，IP【192.168.11.115】，使用UDP节点[分流总节点] [全局代理](TPROXY:1041)代理[2]除1:65535外的所有端口
2023-09-04 01:24:32:   - 备注【server】，IP【192.168.11.243】，使用TCP节点[分流总节点] [全局代理](REDIRECT:1041)代理[2]除51413,18443,55555,6580,58769外的所有端口
2023-09-04 01:24:32:   - 备注【server】，IP【192.168.11.243】，使用UDP节点[分流总节点] [全局代理](TPROXY:1041)代理[2]除1:65535外的所有端口
2023-09-04 01:24:33:   - 备注【pc＆phone】，IP【192.168.11.118】，使用TCP节点[分流总节点] [全局代理](REDIRECT:1041)代理[2]除51413,18443,55555,6580,58769外的所有端口
2023-09-04 01:24:33:   - 备注【pc＆phone】，IP【192.168.11.118】，使用UDP节点[分流总节点] [全局代理](TPROXY:1041)代理[2]除1:65535外的所有端口
2023-09-04 01:24:33:   - 备注【pc＆phone】，IP【192.168.11.103】，使用TCP节点[分流总节点] [全局代理](REDIRECT:1041)代理[2]除51413,18443,55555,6580,58769外的所有端口
2023-09-04 01:24:33:   - 备注【pc＆phone】，IP【192.168.11.103】，使用UDP节点[分流总节点] [全局代理](TPROXY:1041)代理[2]除1:65535外的所有端口
2023-09-04 01:24:33:   - 备注【pc＆phone】，IP【192.168.11.132】，使用TCP节点[分流总节点] [全局代理](REDIRECT:1041)代理[2]除51413,18443,55555,6580,58769外的所有端口
2023-09-04 01:24:33:   - 备注【pc＆phone】，IP【192.168.11.132】，使用UDP节点[分流总节点] [全局代理](TPROXY:1041)代理[2]除1:65535外的所有端口
2023-09-04 01:24:33: TCP默认代理：使用节点[分流总节点] [全局代理](REDIRECT:1041)代理除51413,18443,55555,6580,58769外的所有端口
2023-09-04 01:24:33: UDP默认代理：使用节点[分流总节点] [全局代理](TPROXY:1041)代理除1:65535外的所有端口
2023-09-04 01:24:33: 防火墙规则加载完成！
2023-09-04 01:24:36: 重启 dnsmasq 服务
2023-09-04 01:24:36: 配置定时任务：自动更新【haojie01】订阅。
2023-09-04 01:24:36: 运行完成！

```

**Comment by wizard-leoliao** [comment]
补日志

此为节点日志
> 2023/09/05 00:51:52 127.0.0.1:55553 accepted udp:1.1.1.1:53 [dns-in -> dns-out]
2023/09/05 00:51:52 [Debug] app/dns: domain time.ustc.edu.cn matches following rules: [domain:edu.cn(DNS idx:2) geosite:cn(DNS idx:2)]
2023/09/05 00:51:52 [Debug] app/dns: domain time.ustc.edu.cn will use DNS in order: [UDP:127.0.0.1:15355]
2023/09/05 00:51:52 [Info] app/dns: UDP:127.0.0.1:15355 cache HIT time.ustc.edu.cn -> [] > empty response
2023/09/05 00:51:52 [Info] app/dns: failed to lookup ip for domain time.ustc.edu.cn at server UDP:127.0.0.1:15355 > empty response
2023/09/05 00:51:53 [Debug] [2912004307] proxy/dokodemo: processing connection from: 127.0.0.1:58597
2023/09/05 00:51:53 [Info] [2912004307] proxy/dokodemo: received request for 127.0.0.1:58597
2023/09/05 00:51:53 [Info] [2912004307] app/dispatcher: taking detour [dns-out] for [udp:1.1.1.1:53]
2023/09/05 00:51:53 [Info] [2912004307] proxy/dns: handling DNS traffic to udp:127.0.0.1:15356
2023/09/05 00:51:53 127.0.0.1:58597 accepted udp:1.1.1.1:53 [dns-in -> dns-out]
2023/09/05 00:51:53 [Debug] app/dns: domain broker.miwifi.com matches following rules: [geosite:cn(DNS idx:2)]
2023/09/05 00:51:53 [Debug] app/dns: domain broker.miwifi.com will use DNS in order: [UDP:127.0.0.1:15355]
2023/09/05 00:51:53 [Debug] app/dns: UDP:127.0.0.1:15355 querying DNS for: broker.miwifi.com.
2023/09/05 00:51:53 [Debug] transport/internet/udp: dispatch request to: udp:127.0.0.1:15355
2023/09/05 00:51:53 [Info] app/dns: UDP:127.0.0.1:15355 got answer: broker.miwifi.com. TypeAAAA -> [[240e:f7:c010:102:3::3f8] [240e:947:4001:0:3::3f9] [240e:945:1:6:3::3e8] [240e:95c:803:2:3::3da] [240e:946:6002:207:3::3e3]] 3.613885ms
2023/09/05 00:51:53 [Debug] app/dns: UDP:127.0.0.1:15355 updating IP records for domain:broker.miwifi.com.
2023/09/05 00:51:53 [Debug] [2564681130] proxy/dokodemo: processing connection from: 127.0.0.1:45228
2023/09/05 00:51:53 [Info] [2564681130] proxy/dokodemo: received request for 127.0.0.1:45228
2023/09/05 00:51:53 [Info] [2564681130] app/dispatcher: taking detour [dns-out] for [udp:1.1.1.1:53]
2023/09/05 00:51:53 [Info] [2564681130] proxy/dns: handling DNS traffic to udp:127.0.0.1:15356
2023/09/05 00:51:53 127.0.0.1:45228 accepted udp:1.1.1.1:53 [dns-in -> dns-out]
2023/09/05 00:51:53 [Debug] app/dns: domain broker.miwifi.com matches following rules: [geosite:cn(DNS idx:2)]
2023/09/05 00:51:53 [Debug] app/dns: domain broker.miwifi.com will use DNS in order: [UDP:127.0.0.1:15355]
2023/09/05 00:51:53 [Debug] app/dns: UDP:127.0.0.1:15355 querying DNS for: broker.miwifi.com.
2023/09/05 00:51:53 [Debug] transport/internet/udp: dispatch request to: udp:127.0.0.1:15355
2023/09/05 00:51:53 [Info] app/dns: UDP:127.0.0.1:15355 got answer: broker.miwifi.com. TypeA -> [111.170.130.234 119.96.33.219 119.96.33.217 119.96.33.218 119.96.33.216 111.170.130.228 119.96.33.222 111.170.130.231 119.96.33.215 119.96.33.220 111.170.130.232 111.170.130.230] 3.646382ms
2023/09/05 00:51:53 [Debug] app/dns: UDP:127.0.0.1:15355 updating IP records for domain:broker.miwifi.com.
2023/09/05 00:52:02 [Debug] [2585089649] proxy/dokodemo: processing connection from: 127.0.0.1:57906
2023/09/05 00:52:02 [Info] [2585089649] proxy/dokodemo: received request for 127.0.0.1:57906
2023/09/05 00:52:02 [Info] [2585089649] app/dispatcher: taking detour [dns-out] for [udp:1.1.1.1:53]
2023/09/05 00:52:02 [Info] [2585089649] proxy/dns: handling DNS traffic to udp:127.0.0.1:15356
2023/09/05 00:52:02 127.0.0.1:57906 accepted udp:1.1.1.1:53 [dns-in -> dns-out]
2023/09/05 00:52:32 [Debug] [4175017880] proxy/dokodemo: processing connection from: 127.0.0.1:35867
2023/09/05 00:52:32 [Info] [4175017880] proxy/dokodemo: received request for 127.0.0.1:35867
2023/09/05 00:52:32 [Info] [4175017880] app/dispatcher: taking detour [dns-out] for [udp:1.1.1.1:53]
2023/09/05 00:52:32 [Info] [4175017880] proxy/dns: handling DNS traffic to udp:127.0.0.1:15356
2023/09/05 00:52:32 127.0.0.1:35867 accepted udp:1.1.1.1:53 [dns-in -> dns-out]
2023/09/05 00:52:32 [Debug] app/dns: domain ssl.google-analytics.com matches following rules: [geosite:category-ads(DNS idx:1) geosite:google(DNS idx:1)]
2023/09/05 00:52:32 [Debug] app/dns: domain ssl.google-analytics.com will use DNS in order: [UDP:127.0.0.1:15356]
2023/09/05 00:52:32 [Info] app/dns: UDP:127.0.0.1:15356 cache HIT ssl.google-analytics.com -> [] > empty response
2023/09/05 00:52:32 [Info] app/dns: failed to lookup ip for domain ssl.google-analytics.com at server UDP:127.0.0.1:15356 > empty response
2023/09/05 00:53:02 [Debug] [3214980104] proxy/dokodemo: processing connection from: 127.0.0.1:55036
2023/09/05 00:53:02 127.0.0.1:55036 accepted udp:1.1.1.1:53 [dns-in -> dns-out]
2023/09/05 00:53:02 [Info] [3214980104] proxy/dokodemo: received request for 127.0.0.1:55036
2023/09/05 00:53:02 [Info] [3214980104] app/dispatcher: taking detour [dns-out] for [udp:1.1.1.1:53]
2023/09/05 00:53:02 [Info] [3214980104] proxy/dns: handling DNS traffic to udp:127.0.0.1:15356

此为清空ipset后出现的系统日志
> Tue Sep  5 08:55:49 2023 cron.info crond[6852]: crond (busybox 1.36.0) started, log level 8
Tue Sep  5 08:55:52 2023 daemon.info dnsmasq[7962]: started, version 2.88 cache disabled
Tue Sep  5 08:55:52 2023 daemon.info dnsmasq[7962]: compile time options: IPv6 GNU-getopt no-DBus UBus no-i18n no-IDN DHCP DHCPv6 no-Lua TFTP no-conntrack ipset no-nftset no-auth no-cryptohash no-DNSSEC loop-detect inotify dumpfile
Tue Sep  5 08:55:52 2023 daemon.info dnsmasq[7962]: using nameserver 127.0.0.1#15354
Tue Sep  5 08:55:52 2023 daemon.info dnsmasq[7962]: read /etc/hosts - 7 names
Tue Sep  5 08:55:56 2023 cron.info crond[10147]: crond (busybox 1.36.0) started, log level 8
Tue Sep  5 08:56:42 2023 daemon.err uhttpd[5144]: sh: write error: Broken pipe

**Reply by wizard-leoliao** [reply]
补充日志：
此为出现问题后的节点日志(在agdh中发现应该被代理的域名，确认为dns泄露导致的代理失效)
```
2023/09/14 00:27:33 [Debug] [1314200316] proxy/dokodemo: processing connection from: 127.0.0.1:40217
2023/09/14 00:27:33 [Info] [1314200316] proxy/dokodemo: received request for 127.0.0.1:40217
2023/09/14 00:27:33 [Info] [1314200316] app/dispatcher: taking detour [dns-out] for [udp:1.1.1.1:53]
2023/09/14 00:27:33 127.0.0.1:40217 accepted udp:1.1.1.1:53 [dns-in -> dns-out]
2023/09/14 00:27:33 [Info] [1314200316] proxy/dns: handling DNS traffic to udp:127.0.0.1:15356
2023/09/14 00:27:33 [Info] [1955380716] app/dispatcher: taking detour [dns-out] for [udp:1.1.1.1:53]
2023/09/14 00:27:33 [Info] [1955380716] proxy/dns: handling DNS traffic to udp:127.0.0.1:15356
2023/09/14 00:27:33 127.0.0.1:40623 accepted udp:1.1.1.1:53 [dns-in -> dns-out]
2023/09/14 00:27:33 [Debug] app/dns: domain ssl.google-analytics.com matches following rules: [geosite:category-ads-all(DNS idx:1) geosite:google(DNS idx:1)]
2023/09/14 00:27:33 [Debug] app/dns: domain ssl.google-analytics.com will use DNS in order: [UDP:127.0.0.1:15356]
2023/09/14 00:27:33 [Debug] app/dns: domain ssl.google-analytics.com matches following rules: [geosite:category-ads-all(DNS idx:1) geosite:google(DNS idx:1)]
2023/09/14 00:27:33 [Info] app/dns: UDP:127.0.0.1:15356 cache HIT ssl.google-analytics.com -> [] > empty response
2023/09/14 00:27:33 [Debug] app/dns: domain ssl.google-analytics.com will use DNS in order: [UDP:127.0.0.1:15356]
2023/09/14 00:27:33 [Info] app/dns: failed to lookup ip for domain ssl.google-analytics.com at server UDP:127.0.0.1:15356 > empty response
2023/09/14 00:27:33 [Debug] app/dns: UDP:127.0.0.1:15356 cache HIT ssl.google-analytics.com -> [172.217.24.104]
2023/09/14 00:27:33 [Info] [2984924693] app/proxyman/inbound: connection ends > proxy/socks: connection ends > context canceled
```

**Comment by wizard-leoliao** [comment]
相同固件重新安装貌似解决了这个问题......奇怪的问题

**Reply by liu245933567** [reply]
老哥，什么是相同固件重新安装？就是把passwall重新安装一遍？

**Reply by wizard-leoliao** [reply]
重新安装固件。我现在已全面转向使用immortalwrt+homeproxy(内核为singbox)，ui逻辑清晰且非常稳定，但需要学习singbox的配置逻辑。
